### ISSUE TITLE
ng update fails when using local library

### ISSUE THREAD

**Issue by elboukhari** [angular/angular-cli#13059]
Bug Report or Feature Request (mark with an x)
- [X ] bug report -> please search issues before submitting
- [ ] feature request
Command (mark with an x)
- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [x ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
Versions
node --version : v10.13.0
npm --version : 6.4.1
ng --version :

`Angular CLI: 7.1.0
Node: 10.13.0
OS: linux x64
Angular: 7.1.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, http, language-service, material
... material-moment-adapter, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.11.0
@angular-devkit/build-angular     0.11.0
@angular-devkit/build-optimizer   0.11.0
@angular-devkit/build-webpack     0.11.0
@angular-devkit/core              7.1.0
@angular-devkit/schematics        7.1.0
@ngtools/webpack                  7.1.0
@schematics/angular               7.1.0
@schematics/update                0.11.0
rxjs                              6.3.3
typescript                        3.1.6
webpack                           4.26.1`

Repro steps
just type :
ng update

The log given by the failure
Not found : XXXXXX ()

In my package.json this library is declared as :

I'm still having the issue even with :
"@angular/cli": "7.1.0"

the error is the same as in the previous comment:
404 Not Found - GET https://registry.npmjs.org/XXXXXXX - Not found

Desired functionality
just ignore local libraries and update angular ones

Mention any other details that might be useful

**Comment by alan-agius4** [comment]
Hi, when you say local package? is it in a private npm repo? or using the `file://` protocol?

**Comment by elboukhari** [comment]
it's ` file://` protocol, here is the corresponding line in `package.json` : 

`"MY-AWeSOME-LOCAL-LIB": "file:lib/MY-AWeSOME-LOCAL-LIB.tgz",`

**Comment by meta72** [comment]
@alan-agius4 I have the same issue with a private npm repo. When I update any node module it says that it can't find my private module (which is on my private repo) on npmjs.org. This was working until cli version 7.0.7

**Comment by denisoby** [comment]
@alan-agius4 the problem is in angular-cli/packages/schematics/update/update/npm.ts

```
const resultPromise = pacote.packument(
    packageName,
    {
      'full-metadata': true,
      ...npmrc,
      registry: options && options.registryUrl,
    },
  );
```

This line contains error: 
```registry: options && options.registryUrl```

If you pass any cli options, but don't pass registryUrl inside this options - this line will generate "registry: undefined" that always overrides ".npmrc" value.

And pacote when see this - uses default registry.npmjs.org

The fix is simple - check that options.registryUrl has some value, before passing it to pacote.
I can try to make a fix, and also some tests for it :)

**Comment by clydin** [comment]
That’s a different problem.  Local packages; as well as git packages are not suppported by ‘ng update’ currently.

**Comment by elboukhari** [comment]
Even with CLI 7.1.1 I still have the issue. 

Here is my config:
`[08:41:18][~/Documents/devs/my-project/frontend]$ ng version

**Angular CLI: 7.1.1**
Node: 10.13.0
OS: linux x64
Angular: 7.1.1
... animations, cli, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.11.0
@angular-devkit/build-angular      0.11.0
@angular-devkit/build-optimizer    0.11.0
@angular-devkit/build-webpack      0.11.0
@angular-devkit/core               7.1.0
@angular-devkit/schematics         7.1.1
@angular/cdk                       7.1.0
@angular/material                  7.1.0
@angular/material-moment-adapter   7.1.0
@ngtools/webpack                   7.1.0
@schematics/angular                7.1.1
@schematics/update                 0.11.1
rxjs                               6.3.3
typescript                         3.1.6
webpack                            4.26.1
    
[08:41:24][~/Documents/devs/my-project/frontend]$ ng update
**404 Not Found - GET https://registry.npmjs.org/xxx-xxx-xxx  - Not found**
`

**Comment by beleta** [comment]
Same problem here.

`$ ng update`
`404 Not Found - GET https://registry.npmjs.org/private-pkg - Not found`

Where private-pkg is a package stored in a Nexus repository. 

` ng update` worked in the same scenario with angular-cli before 7.1.0.

**Comment by JohnnyDevNull** [comment]
Same problem here.

`C:\lvs\git\my-app [upgrade-core-libs-and-angular +0 ~4 -0 !]> ng update @angular/cli`

Result

`404 Not Found - GET https://registry.npmjs.org/@sk%2fsk-core - Not found`

Where private-pkg is a package stored in a ProGet repository.

ng update worked in the same scenario with angular-cli in Angular 6

**Comment by beleta** [comment]
In latest release, 7.1.3, the problem seems that has been fixed.

**Comment by mgechev** [comment]
@elboukhari would you try 7.1.3 and verify if the issue still persists?

**Comment by elboukhari** [comment]
Hi @mgechev I still have the same issue even under 7.1.3 ! 
@beleta I'm curious how do you declare your local dependency in your package.json ?

```
[08:31:14][~/Documents/devs/my-project/frontend]$ ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 7.1.3
Node: 10.13.0
OS: linux x64
Angular: 7.1.1
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.11.0
@angular-devkit/build-angular      0.11.0
@angular-devkit/build-optimizer    0.11.0
@angular-devkit/build-webpack      0.11.0
@angular-devkit/core               7.1.0
@angular-devkit/schematics         7.1.3
@angular/cdk                       7.1.0
@angular/cli                       7.1.3
@angular/material                  7.1.0
@angular/material-moment-adapter   7.1.0
@ngtools/webpack                   7.1.0
@schematics/angular                7.1.3
@schematics/update                 0.11.3
rxjs                               6.3.3
typescript                         3.1.6
webpack                            4.26.1
    
[08:31:46][~/Documents/devs/my-project/frontend]$ ng update
404 Not Found - GET http://registry.npmjs.org/MY-LIB - Not found
```

And here is the declaration of the lib in my package.json : 

```
...
"dependencies": {
...
"MY-LIB": "file:lib/MY-LIB.tgz",
...
}
...
```

where "MY-LIB" is my local lib

**Comment by JohnnyDevNull** [comment]
For me with ProGet Package the error also still exists with 7.1.3

**Comment by kpaxton** [comment]
I'm getting this on @angular packages with 7.1.3.
I'm on a closed network but have an npm registry mirror set up by my company at a different url.
When I run `ng update` I get `404 Not Found - GET https://registry.npmjs.org/@angular%2fcompiler-cli`.

The package is different each time I run it but it's typically angular packages.

**Comment by beleta** [comment]
> Hi @mgechev I still have the same issue even under 7.1.3 !
> @beleta I'm curious how do you declare your local dependency in your package.json ?
> 
> ```
> [08:31:14][~/Documents/devs/my-project/frontend]$ ng version
> 
>      _                      _                 ____ _     ___
>     / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
>    / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
>   / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
>  /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
>                 |___/
>     
> 
> Angular CLI: 7.1.3
> Node: 10.13.0
> OS: linux x64
> Angular: 7.1.1
> ... animations, common, compiler, compiler-cli, core, forms
> ... http, language-service, platform-browser
> ... platform-browser-dynamic, router
> 
> Package                            Version
> ------------------------------------------------------------
> @angular-devkit/architect          0.11.0
> @angular-devkit/build-angular      0.11.0
> @angular-devkit/build-optimizer    0.11.0
> @angular-devkit/build-webpack      0.11.0
> @angular-devkit/core               7.1.0
> @angular-devkit/schematics         7.1.3
> @angular/cdk                       7.1.0
> @angular/cli                       7.1.3
> @angular/material                  7.1.0
> @angular/material-moment-adapter   7.1.0
> @ngtools/webpack                   7.1.0
> @schematics/angular                7.1.3
> @schematics/update                 0.11.3
> rxjs                               6.3.3
> typescript                         3.1.6
> webpack                            4.26.1
>     
> [08:31:46][~/Documents/devs/my-project/frontend]$ ng update
> 404 Not Found - GET http://registry.npmjs.org/MY-LIB - Not found
> ```
> And here is the declaration of the lib in my package.json :
> 
> ```
> ...
> "dependencies": {
> ...
> "MY-LIB": "file:lib/MY-LIB.tgz",
> ...
> }
> ...
> ```
> where "MY-LIB" is my local lib

I am using a private npm repository and my dependencies are declared this way:

```
  "dependencies": {
    ...
    "carver-login": "0.0.3",
    "carver-validators": "0.0.1",
    ...
  },

```

**Comment by Kurt29** [comment]
This issue still exists in Version @angular/cli 7.3.2

**Comment by nsebs** [comment]
This is still happening with @angular/cli 7.3.6

```bash
$ ng update
Cannot read property '0.7.2' of undefined
```
```bash
$ ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 7.3.6
Node: 8.11.4
OS: linux x64
Angular: 7.2.4
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.0
@angular-devkit/build-angular     0.13.0
@angular-devkit/build-optimizer   0.13.0
@angular-devkit/build-webpack     0.13.0
@angular-devkit/core              7.3.6
@angular-devkit/schematics        7.3.6
@angular/cdk                      7.3.1
@angular/cli                      7.3.6
@angular/material                 7.3.1
@ngtools/webpack                  7.3.0
@schematics/angular               7.3.6
@schematics/update                0.13.6
rxjs                              6.4.0
typescript                        3.2.4
webpack                           4.29.0
```

In `package.json`:
```
  "dependencies": {
    ...
    "my-custom-lib": "./libs/my-custom-lib-v0.7.2.tgz",
    ...
  }
```

**Comment by meta72** [comment]
For me this issue is fixed in version 7.3.9. I didn't test earlier versions. I am using a private npm repository (that had the same issue since version 7.0.7) for some node modules, not a local path.
```
$ ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 7.3.9
Node: 10.15.3
OS: win32 x64
Angular: 7.2.15
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.9
@angular-devkit/build-angular     0.13.9
@angular-devkit/build-optimizer   0.13.9
@angular-devkit/build-webpack     0.13.9
@angular-devkit/core              7.3.9
@angular-devkit/schematics        7.3.9
@angular/cli                      7.3.9
@ngtools/webpack                  7.3.9
@schematics/angular               7.3.9
@schematics/update                0.13.9
rxjs                              6.5.2
typescript                        3.2.4
webpack                           4.29.0
```

**Comment by Kim-Andersen** [comment]
Tested in 8.0.3 and not working.

**Comment by meta72** [comment]
I recently updated from Angular 7.2.15 and angular-cli 7.3.9 to Angular 8.0.0 and angular-cli 8.0.2. I used `ng update @angular/cli @angular/core` and it worked. I then updated to Angular 8.0.3 and angular-cli 8.0.4, which also worked. I then updated some other module, just to see that `ng update` still works with angular-cli 8.0.4.

Disclaimer: I don't have a path to a file for my custom node modules. I have a private repository with my custom modules that also serves all modules from `registry.npmjs.org`.
